src/defrag.c

```diff
@@ -761,16 +756,20 @@ void defragScanCallback(void *privdata, const dictEntry *de) {
     server.stat_active_defrag_scanned++;
 }
 
-static void defragKvstoreDefragScanCallBack(dict **d) {
-    dict *newd;
-    /* handle the dict struct */
-    if ((newd = activeDefragAlloc(*d)))
-        *d = newd;
-    dictDefragTables(*d);
-}
+static void defragPubsubCallback(dict **d) {
+    /* defrag the dict tables */
+    dictDefragTables(d);
 
-void activeDefragKvstore(kvstore *kvs) {
-    kvstoreDictLUTDefrag(kvs, defragKvstoreDefragScanCallBack);
+    /* Defragment the pubsub clients dictionary which is the 'value' part of the dictionary.
+     * We don't defrag the keys because they are also referenced by c->pubsub(shard)_channels. */
+    dictEntry *de;
+    dictIterator *di = dictGetIterator(*d);
+    while((de = dictNext(di)) != NULL) {
+        dict *clients = dictGetVal(de);
+        dictDefragTables(&clients);
+        dictSetVal(*d, de, clients);
+    }
```

Comment:
i'm afraid this loop will take too long to defrag in defragOtherGlobals (can cause a massive latency spike / freeze).
i think we should handle it with scan, like we handle keys.

Reply:
I'm struggling with where to increment defragging the pubsub with scan.
Putting it in `defragOtherGlobals` is too slow, and in the middle of defragging keys feels like waster of cpu, so
i chooise to do it every 16 iterations.
And i add defragmention to clients pubsub, which from testting bringing more benefits than defragging server pubsub.

Reply:
Test python script:
Run this script 3 times, then close the first 2, the fragmentation ratio will be > 10%, if just defragment server->pubsub(shard), the fragmentation ratio won't go down too much, after adding client->pubsub(shard) defragmentation, the fragmentation ratio can go down to < 1%.

```python
import redis
import threading
import random
import string

def random_string(length):
    letters = string.ascii_lowercase
    return ''.join(random.choice(letters) for _ in range(length))

def subscribe_to_channels(redis_conn, channels):
    pubsub = redis_conn.pubsub()
    pubsub.subscribe(channels)
    print("subscribe_to_channels")
    for message in pubsub.listen():
        if message['type'] == 'message':
            print(f"Received message on channel {message['channel']}: {message['data']}")

def create_connections(num_connections, num_channels, channel_length):
    connections = []

    channels = [random_string(channel_length) for _ in range(num_channels)]
    for _ in range(num_connections):
        connection = redis.StrictRedis(host='localhost', port=6379, db=0)
        thread = threading.Thread(target=subscribe_to_channels, args=(connection, channels))
        thread.start()

        connections.append(connection)

    return connections

if __name__ == "__main__":
    num_connections = 1000
    num_channels_per_connection = 5000
    channel_length = 500

    connections = create_connections(num_connections, num_channels_per_connection, channel_length)
    for thread in threading.enumerate():
        if thread != threading.current_thread():
            thread.join()

```